test/Microsoft.Azure.Devices.Edge.Test.Common/config/EdgeConfigBuilder.cs

```diff
@@ -0,0 +1,160 @@
+            List<ModuleConfiguration> modules = this.moduleBuilders
+                .Where(b => b.Key != "$edgeAgent")
+                .Select(b => b.Value.Build())
+                .ToList();
+
+            // Build edge agent
+            modules.Insert(0, this.BuildEdgeAgent(modules));
+
+            // Compose edge configuration
+            var config = new ConfigurationContent
+            {
+                ModulesContent = new Dictionary<string, IDictionary<string, object>>()
+            };
+
+            var moduleNames = new List<string>();
+
+            foreach (ModuleConfiguration module in modules)
+            {
+                moduleNames.Add(module.Name);
+
+                if (module.DesiredProperties.Count != 0)
+                {
+                    config.ModulesContent[module.Name] = new Dictionary<string, object>
+                    {
+                        ["properties.desired"] = module.DesiredProperties
+                    };
+                }
+            }
+
+            return new EdgeConfiguration(this.deviceId, moduleNames, config);
+        }
+
+        ModuleConfiguration BuildEdgeAgent(IEnumerable<ModuleConfiguration> configs)
+        {
+            if (!this.moduleBuilders.TryGetValue("$edgeAgent", out IModuleConfigBuilder agentBuilder))
+            {
+                agentBuilder = this.AddEdgeAgent();
+            }
+
+            // Settings boilerplate
+            var settings = new Dictionary<string, object>()
+            {
+                ["minDockerVersion"] = "v1.25"
+            };
+
+            // Registry credentials under settings
+            this.registry.ForEach(
+                r =>
+                {
+                    settings["registryCredentials"] = new
+                    {
+                        reg1 = new
+                        {
+                            r.username,
+                            r.password,
+                            r.address
+                        }
+                    };
+                });
+
+            // Deployment info for edge agent
+            var systemModules = new Dictionary<string, object>()
+            {
+                // We need to call agentBuilder.Build() *early* here because we
+                // need the agent's own deployment information (e.g. image, env
+                // vars). Even though we aren't finished building the agent
+                // config yet, the agent builder has enough info at this
+                // point to provide what we need. We'll call Build() again
+                // later to get an updated module configuration that includes
+                // deployment info for other modules, plus desired properties
+                // for all modules.
+                [ParseModuleName(agentBuilder.Name).name] = agentBuilder.Build().Deployment
+            };
+
+            // Deployment info for all other modules
+            var modules = new Dictionary<string, object>();
```

Comment:
should it be better to move this block for other module deployment into line147 if block?

Reply:
I don't understand...

This statement is creating the JSON object under `modulesContent.$edgeAgent.properties.desired.systemModules`. The if block at line 147  is creating the JSON object under `modulesContent.$edgeAgent.properties.desired.modules`.

Reply:
sorry for confusion.